nomad/acl_endpoint_test.go

```diff
@@ -1453,85 +1453,175 @@ func TestACLEndpoint_Bootstrap_Reset(t *testing.T) {
 func TestACLEndpoint_UpsertTokens(t *testing.T) {
 	ci.Parallel(t)
 
-	s1, root, cleanupS1 := TestACLServer(t, nil)
-	defer cleanupS1()
-	codec := rpcClient(t, s1)
-	testutil.WaitForLeader(t, s1.RPC)
-
-	// Create the register request
-	p1 := mock.ACLToken()
-	p1.AccessorID = "" // Blank to create
-
-	// Lookup the tokens
-	req := &structs.ACLTokenUpsertRequest{
-		Tokens: []*structs.ACLToken{p1},
-		WriteRequest: structs.WriteRequest{
-			Region:    "global",
-			AuthToken: root.SecretID,
+	// Each sub-test uses the same server to avoid creating a new one for each
+	// test. This means some care has to be taken with resource naming, but
+	// does avoid lots of calls to systems such as freeport.
+	testServer, rootACLToken, testServerCleanup := TestACLServer(t, nil)
+	defer testServerCleanup()
+	codec := rpcClient(t, testServer)
+	testutil.WaitForLeader(t, testServer.RPC)
+
+	testCases := []struct {
+		name   string
+		testFn func(testServer *Server, aclToken *structs.ACLToken)
+	}{
+		{
+			name: "valid client token",
+			testFn: func(testServer *Server, aclToken *structs.ACLToken) {
+
+				// Create the register request with a mocked token. We must set
+				// an empty accessorID, otherwise Nomad treats this as an
+				// update request.
+				p1 := mock.ACLToken()
+				p1.AccessorID = ""
+
+				req := &structs.ACLTokenUpsertRequest{
+					Tokens: []*structs.ACLToken{p1},
+					WriteRequest: structs.WriteRequest{
+						Region:    DefaultRegion,
+						AuthToken: aclToken.SecretID,
+					},
+				}
+				var resp structs.ACLTokenUpsertResponse
+				require.NoError(t, msgpackrpc.CallWithCodec(codec, structs.ACLUpsertTokensRPCMethod, req, &resp))
+				require.Greater(t, resp.Index, uint64(0))
```

Comment:
leave the type casting in the dust
https://github.com/shoenig/test/blob/main/must/must.go#L235